File: 20211223/boj1326/main.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int bfs(int a, int b, vector<int> cost, int n) {
    queue<int> q;
    q.push(a-1);
    vector<int> visited(n, -1);
    visited[a-1] = 0;
    while(!q.empty()) {
        int node = q.front();
        q.pop();
        for(int i = 0; i < n; i++) {
            if((i - node) % cost[node] == 0 && visited[i] == -1) {
                q.push(i);
                visited[i] = visited[node] + 1;
                if(i  == b-1) {
                     return visited[i];
                }
            }
        }
    }
    return -1;
}


int main() {
    int n;
    cin >> n;
    vector<int> cost;
    for(int i = 0; i < n; i++) {
        int temp;
        cin >> temp;
        cost.push_back(temp);
    }
    int a, b;
    cin >> a >> b;
    cout << bfs(a, b, cost, n);
    return 0;
}

File: 20211223/boj1326/main.py
import sys
from collections import deque


def bfs(a, b, cost, n):
    q = deque()
    q.append(a-1)
    visited = [-1] * n
    visited[a-1] = 0
    while q:
        node = q.popleft()
        for i in range(n):
            if (i-node) % cost[node] == 0 and visited[i] == -1:
                q.append(i)
                visited[i] = visited[node] + 1
                if i == b-1:
                    return visited[i]
    return -1


n = int(input())
cost = list(map(int, input().split()))
a, b = map(int, input().split())
cnt = 1
print(bfs(a, b, cost, n))
